### PR TITLE
Projecteuler Lab

### PR DESCRIPTION
**Pull request by weaversa**

### CODE REVIEW

**Comment by weaversa** [comment]
@benjaminjellison created a pull request.

**Comment by weaversa** [comment]
When I try and prove the pythagorean triple from your answers file I get negative numbers in the answer...
```
Cryptol> :prove pythagoreantriple
       pythagoreantriple -1024 1015 1009 = False
```

**Comment by weaversa** [comment]
```
pythagoreantriple : Integer -> Integer -> Integer -> Bit
property pythagoreantriple a b c =
    a^^2 + b^^2 == c^^2 /\
    a + b + c == 1000   /\
    a > 0 /\ b > 0 /\ c > 0
```

Now there's only two:
```
Cryptol> :sat pythagoreantriple
    pythagoreantriple 200 375 425 = True
    pythagoreantriple 375 200 425 = True
```

**Comment by benjaminjellison** [comment]
Yours is more accurate.  I was using `:sat`, not `:prove`, which gives the correct answer.

File: labs/projecteuler/projecteuler.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    x > 1            /\
+    y > 1
+```
+Note that if we don't include the `x != 1` clause we get a trivial factorization.  Now we can use SAT to factor our number:
+```shell
+Main> :sat mult3000013 
+mult3000013 773 3881 = True
+(Total Elapsed Time: 0.684s, using Z3)
+```
+Clearly we don't need SAT solvers to figure out how to subtract one.  And writing simple factorization code in Python factored 3,000,013 in 0.419s.  But there is something really interesting here: to solve the problem, instead of figuring out how to write code which would solve the problem for us, we just had to write the problem in Cryptol code and let a SAT solver solve it for us!  Usually it's much easier to write a problem than write the solution.  We can't always expect the SAT solver to find the answer quickly (this is known as as the [P vs. NP problem](https://en.wikipedia.org/wiki/P_versus_NP_problem)), but there are many times where it will come back with answer quickly.  We like to say that it doesn't cost much to ask a SAT solver.
```
Comment (AssemblyZig) [inline labs/projecteuler/projecteuler.md:42]: ... come back with _an_ answer quickly...

File: labs/projecteuler/projecteuler.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+## Project Euler Problems
+
+[Project Euler](https://projecteuler.net/) is a collection of mathematical problems that can be solved using relatively simple computer code (i.e. to do them, you shouldn't need to create optimized code to run on a supercomputer).  Some of the questions lend themselves to querying a SAT solver.  We've collected them here for you to try your hand at.
```
Comment (AssemblyZig) [inline labs/projecteuler/projecteuler.md:46]: Suggest "... (i.e., the solution shouldn't require optimized code running on an supercomputer)..."

File: labs/projecteuler/projecteuler.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+> 145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145.
+>
+> Find all numbers which are equal to the sum of the factorial of their digits.
+> 
+> Note: as 1! = 1 and 2! = 2 are not sums they are not included.
+>
+> (Aside: these numbers are called [factorions](https://en.wikipedia.org/wiki/Factorion).  It is unclear why Project Euler does not include 1 and 2).
+>
+> Hints: the factorial function is usually defined recursively, but that tends to make SAT solving difficult.  Since you only need to calculate the factorial of the numbers 0-9, make your function just do a case by case calculation.  To get the digital representation of the number, create a function which takes in a number and a list of numbers and returns true exactly when the list is the base 10 representation.  Finally, it can be shown that the most number of di\
+gits a factorion can have is 6.
```
Comment (AssemblyZig) [inline labs/projecteuler/projecteuler.md:67]: remove line break in "digits"

File: labs/projecteuler/projecteuler.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+### [Problem 36](https://projecteuler.net/problem=36)
+
+> The decimal number, 585 = 1001001001_(2) (binary), is palindromic in both bases.
```
Comment (AssemblyZig) [inline labs/projecteuler/projecteuler.md:71]: The representation of the binary is strange.  Perhaps rephrase it as "The decimal number 585 and its binary representation 1001001001 are both [palindromic](https://www.dictionary.com/browse/palindromic)."

File: labs/projecteuler/projecteuler.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+> Let d_(1) be the 1^(st) digit, d_(2) be the 2^(nd) digit, and so on. In this way, we note the following:
+>
+>    * d_(2)d_(3)d_(4)=406 is divisible by 2
+>    * d_(3)d_(4)d_(5)=063 is divisible by 3
+>    * d_(4)d_(5)d_(6)=635 is divisible by 5
+>    * d_(5)d_(6)d_(7)=357 is divisible by 7
+>    * d_(6)d_(7)d_(8)=572 is divisible by 11
+>    * d_(7)d_(8)d_(9)=728 is divisible by 13
+>    * d_(8)d_(9)d_(10)=289 is divisible by 17
+>
```
Comment (AssemblyZig) [inline labs/projecteuler/projecteuler.md:90]: Hmmm... it seems that my markdown browser plugin doesn't interpret the sub-/superscript correctly.  Use the online renderer as suggested by Sean, e.g.: 
> %20=%20M(x)%20\cdot%20x^n%20\\bmod\%20G(x))

**Comment by weaversa** [comment]
@benjaminjellison 🤦‍♂️

File: labs/projecteuler/projecteuler.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+### [Problem 52](https://projecteuler.net/problem=52)
+
+>It can be seen that the number, 125874, and its double, 251748, contain exactly the same digits, but in a different order.
+>
+> Find the smallest positive integer, x, such that 2x, 3x, 4x, 5x, and 6x, contain the same digits. 
+
+### [Problem 59](https://projecteuler.net/problem=59)
+
+> Each character on a computer is assigned a unique code and the preferred standard is ASCII (American Standard Code for Information Interchange). For example, uppercase A = 65, asterisk (*) = 42, and lowercase k = 107.
```
Comment (AssemblyZig) [inline labs/projecteuler/projecteuler.md:107]: Make ASCII definition a link

File: labs/projecteuler/projecteuler.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+> A common security method used for online banking is to ask the user for three random characters from a passcode. For example, if the passcode was 531278, they may asked for the 2nd, 3rd, and 5th characters; the expected reply would be: 317.
+>
+> The text file, keylog.cry, contains fifty successful login attempts.
+>
+> Given that the three characters are always asked for in order, analyse the file so as to determine the shortest possible secret passcode of >unknown length.
+
+### Project Sean
+Sean Weaver (Super Genius ala Wile E Coyote) has a problem in a similar vein:
+
+> Find a four-digit number (greater than 999 and less than 10000) such that the least significant four digits of the square is that number.
```
Comment (AssemblyZig) [inline labs/projecteuler/projecteuler.md:129]: "...four-digit **decimal** number..."  (for the pedants in the crowd)

File: labs/projecteuler/projecteuler.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
+# Properties in Cryptol, or How I Learned to Stop Worrying and Love SAT Solvers
+
+You have learned in the other labs that you can use Cryptol to find answers to problems you coded using the `:sat` command.  As a refresher, let's look at a simple example.  Suppose we want to find the integer that is one less than 1001 (as we said, it's very simple).  We do this by creating a function and a `property`:
+
+```
```
Comment (AssemblyZig) [inline labs/projecteuler/projecteuler.md:5]: The haskell mode provides some highlighting for example code blocks like this.

File: labs/projecteuler/projecteuler_answers.md [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
...he Gospel of John, chapter 1) 1 In the beginning the Word already existed. He was with God, and he was God. 2 He was in the beginning with God. 3 He created everything there is. Nothing exists that he didn't make. 4 Life itself was in him, and this life gives light to everyone. 5 The light shines through the darkness, and the darkness can never extinguish it. 6 God sent John the Baptist 7 to tell everyone about the light so that everyone might believe because of his testimony. 8 John himself was not the light; he was only a witness to the light. 9 The one who is the true light, who gives light to everyone, was going to come into the world. 10 But although the world was made through him, the world didn't recognize him when he came. 11 Even in his own land and among his own people, he was not accepted. 12 But to all who believed him and accepted him, he gave the right to become children of God. 13 They are reborn! This is not a physical birth resulting from human passion or plan, this rebirth comes from God.14 So the Word became human and lived here on earth among us. He was full of unfailing love and faithfulness. And we have seen his glory, the glory of the only Son of the Father."
```
Comment (AssemblyZig) [inline labs/projecteuler/projecteuler_answers.md:345]: Hmmm... is incorporating a bible verse the best choice?  Seems platform-y...

File: labs/projecteuler/projecteuler.md (hunk 7 above)
Comment (benjaminjellison) [reply labs/projecteuler/projecteuler.md:120]: I'm keeping it as is.  I'm willing to be outvoted.

File: labs/projecteuler/projecteuler.md (hunk 8 above)
Comment (weaversa) [reply labs/projecteuler/projecteuler.md:5]: Cryptol's literate system only accepts code blocks that start and end with "clean" backticks.

Comment (weaversa) [reply labs/projecteuler/projecteuler.md:5]: That's how we can have shell commands and literate cryptol in the same file :-)

File: labs/projecteuler/projecteuler.md (hunk 5 above)
Comment (benjaminjellison) [reply labs/projecteuler/projecteuler.md:81]: I'm not sure the renderer gives that much value for how difficult it is to use.

**Pull request merged**